### ISSUE TITLE
Write an Adapter class for ChainableStack?

### ISSUE THREAD

**Issue by SteveBronder** [stan-dev/math#1272]
## Description

Right now when we call the ChainableStack we have to call stuff like

```cpp
ChainableStack::instance().memalloc_.alloc(length_ * sizeof(vari*)));
```

Would there be a big overhead if we called the above class ChainableStackImpl and wrote a little adapter class that would let us call

```cpp
ChainableStack.alloc<vari>(length_ );
```

I see we also have

```cpp
template <int R, int C>
vari** build_vari_array(const Eigen::Matrix<var, R, C>& x) {
  vari** x_vi_
      = ChainableStack::instance().memalloc_.alloc_array<vari*>(x.size());
  for (int i = 0; i < x.size(); ++i) {
    x_vi_[i] = x(i).vi_;
  }
  return x_vi_;
}
```

Instead of writing a whole adapter class could we also just template the above with a type T to make it general?

#### Current Version:
v2.19.1

**Comment by wds15** [comment]
Makes a lot of sense and maybe this can be turned into an actual instance of the stack which has all the methods in it (like recover memory, etc.)?

However, you will need to be very careful if you introduce an adapter. That is, in some places you should probably not use such an adapter (like in the var class). The reason is that the stack access is in some places very performance sensitive and you cannot rely on the compiler optimizing away everything even though you think that the compiler should do that.

**Comment by SteveBronder** [comment]
Yeah we'd have to be pretty clever. I'll think about this during this week or next. I think we could use CRTP

**Comment by bob-carpenter** [comment]
@SteveBronder It'd be helpful if you could be specific about the signatures you're proposing for helper functions.

>  template the above with a type T

I'm not sure what the above is here.  We can't replace the `var`, because the code is `var`-specific. 

>  Would there be a big overhead if we called the above class ChainableStackImpl and wrote a little adapter class that would let us call `ChainableStack.alloc<vari>(length_ );`

That's exactly what `alloc_array` does.  It's used in the body of the code snippet in the original issue.  Specifically,

```
ChainableStack::instance().memalloc_.alloc(length_ * sizeof(vari*)));
```

should be replaced with

```
ChainableStack::instance().memalloc_.alloc_array<vari>(length_);
```

It also takes care of the return pointer type, again as you can see the body.  

Now if you want, we could have one that returns something with more structure than a raw memory array.  But we're usually storing these allocated items as raw memory because we can't use RAII types like `std::vector` or `Eigen::Matrix` in our `vari` implementations.

**Comment by SteveBronder** [comment]
> @SteveBronder It'd be helpful if you could be specific about the signatures you're proposing for helper functions.

I'll look over the current methods we use today so I can put up something more formal.

My main intent is simply that the below is a lot to call

```
ChainableStack::instance().memalloc_.alloc_array<vari>(length_);
```

and it would be nice to be able to call something like `ChainableStack.alloc_array<vari>(length_)`

> Now if you want, we could have one that returns something with more structure than a raw memory array. But we're usually storing these allocated items as raw memory because we can't use RAII types like std::vector or Eigen::Matrix in our vari implementations.

I think what your saying here is more of what I'm shooting for. Why can't we use RAII types? It feels like their must be a semi-reasonable scheme to write that would let us use standard containers. Facebook has an arena allocator that conforms to the ++20 allocator concept which might be useful inspiration

https://github.com/facebook/folly/blob/master/folly/memory/Arena.h#L61

**Comment by wds15** [comment]
Yes! Turning the AD stack into a usual object which has normal RAII style resource allocation would be  much along the lines of my thinking in the parallel design RFC. What I ended up proposing is a compromise between where we are and what I wanted to do. The upside of having the AD stack in an object is that then we can have multiple of these - on per "thing" to do. My interest in that was to be able to compute slices of our work in different contexts and then combine them again.

This global AD stack did make certainly sense at the time, but (I think) we should be moving away from that.

I will have a look at the allocator.

**Comment by bob-carpenter** [comment]
> is a lot to call `ChainableStack::instance().memalloc_.alloc_array<vari>(length_);`

Ah, I see.  I thought you were objecting to all the casting and sizing.  We could 

1. define some utility static class functions in`ChainableStack` and make that 
```
ChainableStack::alloc_array<vari>(size_t)
```

2.  deifne a simple untility function `alloc_array` 
```
template <typename T>
T* alloc_array<vari>(size_t);
```

Either way, the symbols `ChainableStack` and `alloc_array` are in namespace `stan::memory`, I think.

**Comment by bob-carpenter** [comment]
We already allow RAII with a special constructor for vari.  I would like to get rid of the last vestiges of that usage (some remaining linear algebra solves) and completely prohibit vari from malloc.  We can manage memory much more efficiently ourselves.

Having separate AD stacks is a separate issue (no pun intended).

**Comment by SteveBronder** [comment]
> Yes! Turning the AD stack into a usual object which has normal RAII style resource allocation would be much along the lines of my thinking in the parallel design RFC. 

Did you talk about this in the parallel design rfc? I might have not seen it but will def read that over if so

> What I ended up proposing is a compromise between where we are and what I wanted to do. The upside of having the AD stack in an object is that then we can have multiple of these - on per "thing" to do. My interest in that was to be able to compute slices of our work in different contexts and then combine them again.

I definitely agree with the above, I like this line but also think we should move this part of the convo over to discourse since the scope of this issue is essentially just helper functions. 

> This global AD stack did make certainly sense at the time, but (I think) we should be moving away from that.

I think it still makes sense, think about how hoard has the global memory that it holds and then gives out to the sub memory pools.

> I will have a look at the allocator.

Same! Let's both try to read over their stuff then chat later in the week and see if we want to make a more formal proposal

**Comment by SteveBronder** [comment]
Bob yeah I think 1 and 2 are v reasonable (and already sort of exist). I think that's scope I want in this so yeah I'll give those a go! Should be simple.

> We already allow RAII with a special constructor for vari.  I would like to get rid of the last vestiges of that usage (some remaining linear algebra solves) and completely prohibit vari from malloc.  We can manage memory much more efficiently ourselves.

I think I'm using the term RAII wrong. Let me rephrase. I want to be able to use std::vectors with our own allocator. So I want the first half of RAII, acquire memory at construction. But I don't want us to be free'ing memory all loosey goosey.

> Having separate AD stacks is a separate issue (no pun intended).

Agree we can have a discourse post about that
